### PR TITLE
feat: add support for optional external auth providers

### LINKED ISSUE #10109: Make external auth providers optional
## Background

We have a `coder_external_auth` data source that requires the user auth with git or another sevice before creating the template.

## Current behavior

When git auth is added to a template, it is currently required. However, some organizations are migrating from one git provider to another or rolling out a provider to others. This means not all template users have an account on a given service. Terraform is also capable of conditional logic, such as cloning a repo if the user authed

## Proposed behavior

A provider can be marked as optional.

```hcl
data "coder_external_auth" "experimental_github" {
  id = "experimental_github"
  required = false # default is true to avoid a breaking change
}

resource "coder_agent" "my script" {
  # ...
  startup_script=<<EOF
  # this is pseudocode 
  if [${data.coder_parameter.external_auth.token == !null}] then;
    git clone ...
  fi
  EOF
}
```

### PR DESCRIPTION
**Pull request by aslilac**
Screenshot 2024-02-14 at 4 34 01 PM

(screenshot includes frontend changes that will be landing in a separate pull request, because this one is already big enough 😄)

Part of #10109

- Migrate `external_auth_providers` column in the database, in order to be able to contain metadata about providers, such as settings/attributes defined in Terraform.
- Add support for `external_auth_provider.optional` in the Terraform provisioner and provisionerd
- Add `optional` to `TemplateVersionExternalAuth`
  - and fill it with new database goodness when querying `/templateversions/{templateversion}/external-auth`

#### TODO

- who uses `proto.CurrentVersion`? there are some tests failing because of this, and I'm not really sure what to do about it
  - also it doesn't seem like there's any machinery for checking the terraform provider version? kind of unfortunate that you could just start using 0.16 with an old version of coder and `optional` would just silently disappear

### CODE REVIEW

File: provisionerd/runner/runner.go [hunk 1]
```diff
@@ -629,7 +629,7 @@ func (r *Runner) runTemplateImportParse(ctx context.Context) (
 type templateImportProvision struct {
 	Resources             []*sdkproto.Resource
 	Parameters            []*sdkproto.RichParameter
-	ExternalAuthProviders []string
+	ExternalAuthProviders []*sdkproto.ExternalAuthProviderResource
```
Comment (spikecurtis) [inline provisionerd/runner/runner.go:632]: when you convert this type to `proto.CompletedJob_TemplateImport` on line 570, you also need to set the ExternalAuthProviderNames for compatibility with back-level Coderd.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 2]
```diff
@@ -1133,23 +1139,42 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
...
+		if len(externalAuthProviders) == 0 && namesLen > 0 {
+			externalAuthProviders = make([]*sdkproto.ExternalAuthProviderResource, 0, namesLen)
+			for _, providerID := range jobType.TemplateImport.ExternalAuthProvidersNames {
+				externalAuthProviders = append(externalAuthProviders, &sdkproto.ExternalAuthProviderResource{
+					Id: providerID,
+				})
+			}
+		}
+
+		externalAuthProvidersMessage, err := json.Marshal(externalAuthProviders)
```
Comment (spikecurtis) [inline coderd/provisionerdserver/provisionerdserver.go:1170]: I'm a bit wary of directly serializing an `sdkproto` type to the database, as we don't version that type and kind of treat it as something we can change without consequence.

I think you should define the datatype for `ExternalAuthProviders` column in the `database` package (maybe in `models.go`?) and explicitly convert the `sdkproto` type.

Comment (aslilac) [reply coderd/provisionerdserver/provisionerdserver.go:1170]: models.go is generated by sqlc right? would types.go be ok instead?

Comment (aslilac) [reply coderd/provisionerdserver/provisionerdserver.go:1177]: I think that's a good idea tho, also makes it clearer when the value came from a protobuf message vs when we read it from the database.

**Comment by kylecarbs** [comment]
Regarding this UX, should we revert to the old UI? The "Connect" button with "Optional" and then also presumably a "Skip" feels like a lot.

File: coderd/provisionerdserver/provisionerdserver.go (hunk 2 above)
Comment (spikecurtis) [reply coderd/provisionerdserver/provisionerdserver.go:1177]: yeah, types.go is all good; I can never remember which files are generated

File: provisionerd/proto/version.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import (
+	"github.com/coder/coder/v2/coderd/util/apiversion"
+)
+
+const (
+	CurrentMajor = 1
+	CurrentMinor = 1
+)
+
+var CurrentVersion = apiversion.New(CurrentMajor, CurrentMinor)
```
[the remark this answers is not in the document]
Comment (aslilac) [reply provisionerd/proto/version.go:12]: no, this is supposed to be a separate version. @spikecurtis and I added it while pairing, but I'll admit I'm not really sure how the versioning here should work exactly. 😅

[the remark this answers is not in the document]
Comment (johnstcn) [reply provisionerd/proto/version.go:12]: OK 🤔 I don't really understand why we need two versions @spikecurtis?

File: provisioner/terraform/resources.go [hunk 4]
```diff
@@ -699,23 +699,33 @@ func ConvertState(modules []*tfjson.StateModule, rawGraph string) (*State, error
...
 			id, ok := resource.AttributeValues["id"].(string)
 			if !ok {
 				return nil, xerrors.Errorf("external auth id is not a string")
 			}
-			externalAuthProvidersMap[id] = struct{}{}
+			optional := false
+			optionalAttribute, ok := resource.AttributeValues["optional"].(bool)
+			if ok {
+				optional = optionalAttribute
+			}
```
[the remark this answers is not in the document]
Comment (aslilac) [reply provisioner/terraform/resources.go:718]: the terraform-provider will always make sure this is a `bool`, and I can't ever imagine us changing the type of this attribute. 🤔

File: coderd/database/migrations/000194_external_auth_providers_jsonb.up.sql [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+DECLARE
+  result jsonb;
+BEGIN
+  SELECT
+    jsonb_agg(jsonb_build_object('id', value::text)) INTO result
+  FROM
+    unnest($1) AS value;
+  RETURN result;
+END;
+$$;
```
[the remark this answers is not in the document]
Comment (aslilac) [reply coderd/database/migrations/000194_external_auth_providers_jsonb.up.sql:20]: nope!

File: provisionerd/proto/version.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (spikecurtis) [reply provisionerd/proto/version.go:12]: There only needs to be one version, and it should be `provisionerd` not `provisionersdk`.

There are 2 distinct, but related protocols: 

`provisionersdk/proto` is  Provisionerd <-> Provisioner (terraform/echo) 

`provisionerd/proto` is Provisionerd <-> Coderd (provisionerdserver subcomponent)

Right now the former is only ever spoken in-memory. Eventually we want it to be spoken between K8s jobs and the Provisionerd, but that's not been built.  So, it doesn't need a protocol version.

The latter, is spoken over the wire, and needs a protocol version.

I think what happened, is when @johnstcn introduced the APIVersion in #11393 he mistakenly added it to `provisionersdk` and I didn't catch it in code review.  It should be in `provisionerd`. Notice that in #12191 , after we check the version, we start a service defined in `provisionerd/proto`.

[the remark this answers is not in the document]
Comment (johnstcn) [reply provisionerd/proto/version.go:12]: Ah gotcha, I'll open a separate PR to move this to the right place.

[the remark this answers is not in the document]
Comment (johnstcn) [reply provisionerd/proto/version.go:12]: https://github.com/coder/coder/pull/12225

[the remark this answers is not in the document]
Comment (aslilac) [reply provisionerd/proto/version.go]: ok, I've merged those changes into this branch now!

File: provisionerd/proto/version.go [hunk 6]
```diff
@@ -7,7 +7,7 @@ const (
 	CurrentMinor = 0
 )
 
-// VersionCurrent is the current provisionerd API version.
+// CurrentVersion is the current provisionerd API version.
 // Breaking changes to the provisionerd API **MUST** increment
 // CurrentMajor above.
-var VersionCurrent = apiversion.New(CurrentMajor, CurrentMinor)
+var CurrentVersion = apiversion.New(CurrentMajor, CurrentMinor)
```
Comment (aslilac) [inline provisionerd/proto/version.go:13]: I'm renaming this for consistency with other similar `apiversion` declarations

File: coderd/database/dump.sql [hunk 7]
```diff
@@ -880,8 +880,6 @@ CREATE VIEW template_version_with_user AS
    FROM (public.template_versions
      LEFT JOIN visible_users ON ((template_versions.created_by = visible_users.id)));
 
-COMMENT ON VIEW template_version_with_user IS 'Joins in the username + avatar url of the created by user.';
-
```
Comment (johnstcn) [inline coderd/database/dump.sql:884]: nit: don't forget to re-add the comment on the view in the up and down migrations

**Review by johnstcn — APPROVED** [review]

File: coderd/database/dump.sql (hunk 7 above)
Comment (aslilac) [reply coderd/database/dump.sql:884]: good catch, thank you!

**Pull request merged**